osu.Framework/Utils/OrderAttribute.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace osu.Framework.Utils
+{
+    public static class OrderAttributeUtils
+    {
+        /// <summary>
+        /// Get values of an enum in order. Supports custom ordering via <see cref="OrderAttribute"/>.
+        /// </summary>
+        public static IEnumerable<T> GetValuesInOrder<T>()
+        {
+            var type = typeof(T);
+
+            if (!type.IsEnum)
+                throw new InvalidOperationException("T must be an enum");
+
+            IEnumerable<T> items = (T[])Enum.GetValues(type);
+
+            return GetValuesInOrder(items);
+        }
+
+        /// <summary>
+        /// Get values of a collection of enum values in order. Supports custom ordering via <see cref="OrderAttribute"/>.
+        /// </summary>
+        public static IEnumerable<T> GetValuesInOrder<T>(IEnumerable<T> items)
+        {
+            var type = typeof(T);
+
+            if (!type.IsEnum)
+                throw new InvalidOperationException("T must be an enum");
+
+            if (!(Attribute.GetCustomAttribute(type, typeof(HasOrderedElementsAttribute)) is HasOrderedElementsAttribute orderedAttr))
+                return items;
+
+            return items.OrderBy(i =>
+            {
+                if (type.GetField(i.ToString())?.GetCustomAttributes(typeof(OrderAttribute), false).FirstOrDefault() is OrderAttribute attr)
+                    return attr.Order;
+
+                if (orderedAttr.AllowPartialOrdering)
+                    return (int)Enum.Parse(type, i.ToString());
+
+                throw new ArgumentException($"Not all values of {nameof(T)} have {nameof(OrderAttribute)} specified.");
```

Comment:
```suggestion
                throw new ArgumentException($"Not all values of {typeof(T)} have {nameof(OrderAttribute)} specified.");
```

`nameof` is compile-time and does not work with generics